### PR TITLE
More Linux 4.4 issues

### PR DESCRIPTION
**Pull request by pallas**

### CODE REVIEW

File: linuxmodule/click-linuxtool.pl [hunk 1]
```diff
@@ -343,6 +343,10 @@ ($$)
 	    if ($d eq "irq.h") {
 	    	s{enum irqchip_irq_state}{int}g;
 	    }
+	    if ($d eq "atomic.h") {
+	        s{\batomic_and\b}{atomic_and_value}g;
+	        s{\batomic64_and\b}{atomic64_and_value}g;
+	    }
```
Comment (kohler) [inline linuxmodule/click-linuxtool.pl]: This isn't the right fix.

\band\b should match ONLY and on its own.

\b is supposed to match boundaries between \w and \W, and \w includes underscore.

Comment (kohler) [inline linuxmodule/click-linuxtool.pl]: I see, the symbol `atomic_and` is being built up by token pasting...

Comment (kohler) [inline linuxmodule/click-linuxtool.pl]: See 274beba4cf34de7359501e6bca28cdfa190c3bef

File: linuxmodule/click-linuxtool.pl [hunk 2]
```diff
@@ -414,6 +414,7 @@ ($$)
 	    }
 	    if ($d eq "compiler.h" || $d eq "linkage.h") {
 		s<^#define ACCESS_ONCE\(x\) \(\*\(volatile typeof\(x\) \*\)\&\(x\)\)><#define ACCESS_ONCE(x) (*(typeof(x) * volatile)&(x))>m;
+		s<^#define ACCESS_ONCE\(x\) \(\*__ACCESS_ONCE\(x\)\)><#define ACCESS_ONCE(x) (*(typeof(x) * volatile)&(x))>m;
```
Comment (kohler) [inline linuxmodule/click-linuxtool.pl]: This just redefines `ACCESS_ONCE` to what it used to be.

File: linuxmodule/click-linuxtool.pl [hunk 3]
```diff
@@ -340,6 +340,9 @@ ($$)
 	    if ($d eq "rbtree_latch.h") {
 	    	s{container_of\(node, struct latch_tree_node, node\[idx\]\)}{idx==0 ? container_of(node, struct latch_tree_node, node[0]) : container_of(node, struct latch_tree_node, node[1])};
 	    }
+	    if ($d eq "irq.h") {
+	    	s{enum irqchip_irq_state}{int}g;
+	    }
```
Comment (kohler) [inline linuxmodule/click-linuxtool.pl]: I fixed this by fixing its type with `enum irqchip_irq_state : int;`.

But that requires C++11. Your fix is probably preferable…

File: linuxmodule/click-linuxtool.pl [hunk 4]
```diff
@@ -337,6 +337,9 @@ ($$)
 	    s{(\w+)\s*-\s*\(\s*void\s*\*\s*\)}{(uintptr_t)$1 - (uintptr_t)}g;
 
 	    # stuff for particular files (what a shame)
+	    if ($d eq "rbtree_latch.h") {
+	    	s{container_of\(node, struct latch_tree_node, node\[idx\]\)}{idx==0 ? container_of(node, struct latch_tree_node, node[0]) : container_of(node, struct latch_tree_node, node[1])};
+	    }
```
Comment (kohler) [inline linuxmodule/click-linuxtool.pl]: Which GCC are you using? gcc-5.2.1 does not complain about the unmodified code.

**Comment by kohler** [comment]
523985e: i fixed this a different way

656c5f1: reverted my `enum irq_... : int` in favor of something like this

5466984: fixed this a different way (ensuring that we declare atomic_and)

e110b22: Is this change actually needed? With gcc-5.2.1, on my recent ubuntu VM, i don't seem to need the `ACCESS_ONCE` change, and in recent commits, have commented it out. If we still need this change, perhaps something like `typeof(x + 0)` will help. Can you copy/paste the error message you're getting?

**Comment by kohler** [comment]
So as of 7d0d6247d52dab07af2ebde171328a6891dee091 I have a version that compiles and installs on an x86-64 4.2.0 kernel. I haven't tested packet forwarding.

**Comment by pallas** [comment]
@kohler I am still getting `/home/pallas/click/include/click-linuxmodule/include3/linux/atomic.h:573:20: error: 'atomic64_and' was not declared in this scope` but not the atomic_and version on kohler/master.

**Comment by kohler** [comment]
I think 783865a013e36b2649e5429fdbe9a160aa28041d should fix that?

**Comment by kohler** [comment]
It didn't, but now 8418a77 does compile, and a linuxmodule installs and runs conf/test.click, on ubuntu 16.04/linux 4.4.

**Comment by pallas** [comment]
Yep, I can confirm that it configures. Compiling now.

**Comment by pallas** [comment]
Great success! Thanks for the help!!

**Pull request closed without merging**